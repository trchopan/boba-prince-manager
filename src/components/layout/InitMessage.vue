<template>
  <div id="initial-loading">
    <img src="/images/bloody-smile.jpg" alt="initializing image" />
    <p>...Let's put a smile on that face</p>
  </div>  
</template>

<script>
export default {
  name: "InitMessage"
};
</script>


<style lang="scss" scoped>
#initial-loading {
  font-family: Roboto;
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  img {
    width: 6rem;
    height: auto;
    border-radius: 50%;
  }
}
</style>
