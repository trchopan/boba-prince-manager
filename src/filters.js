import Vue from "vue";

export const filters = {
  titleCase: str =>
    str
      .toLowerCase()
      .split(" ")
      .map(function(word) {
        return word.replace(word[0], word[0].toUpperCase());
      })
      .join(" ")
};

Object.keys(filters).forEach(key => Vue.filter(key, filters[key]));
