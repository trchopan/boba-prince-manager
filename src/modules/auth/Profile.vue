<template>
  <v-layout column>
    <my-header>{{ $t.greeting }}</my-header>
    <v-layout row align-center justify-center>
      <v-flex v-if="!profile" xs12 sm8 md6>
        <p class="text-xs-center">{{ $t.profileIsBeingCreated }}</p>
      </v-flex>
      <v-flex v-else xs12 sm8 md5>
        <transition name="fade" mode="out-in">
          <v-card v-if="!editMode" key="profile-display" class="elevation-1">
            <v-list>
              <v-list-tile avatar>
                <v-list-tile-avatar>
                  <img :src="profile.avatar" alt="big avatar">
                </v-list-tile-avatar>
                <v-list-tile-content>
                  <v-list-tile-title>{{ profile.fullname }}</v-list-tile-title>
                  <v-list-tile-sub-title>{{ authUser.email }}</v-list-tile-sub-title>
                </v-list-tile-content>
                <v-list-tile-action>
                  <v-btn outline @click="editMode = true">{{ $t.edit }}</v-btn>
                </v-list-tile-action>
              </v-list-tile>
            </v-list>
          </v-card>
          <v-card
            v-if="editMode"
            key="profile-edit"
            class="elevation-6 pt-3 px-3 text-xs-center"
          >
            <v-avatar size="80">
              <img :src="profile.avatar" alt="big avatar">
            </v-avatar>
            <v-text-field
              id="fullname"
              prepend-icon="person"
              :label="$t.fullname"
              type="text"
              v-model="fullname"
              required
            />
            <v-text-field
              id="avatar"
              prepend-icon="lock"
              :label="$t.avatar"
              type="text"
              v-model="avatar"
              required
            />
            <v-card-actions>
              <v-spacer/>
              <template v-if="!prestine">
                <v-icon
                  class="mr-1"
                  color="success"
                >{{ loading.profile ? 'hourglass_empty' : 'check_circle_outline' }}</v-icon>
                <span class="mr-2">{{ loading.profile ? $t.loading : $t.saved }}</span>
              </template>
              <v-btn
                outline
                color="primary"
                type="button"
                :disabled="loading.profile"
                @click="back()"
              >{{ $t.back }}</v-btn>
            </v-card-actions>
          </v-card>
        </transition>
      </v-flex>
    </v-layout>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import { debounce } from "@/helpers";

const languagesMap = {
  greeting: { vi: "Xin chào, bạn phẻ hok?", en: "Hi, how are you today?" },
  profileIsBeingCreated: {
    vi: "Tài khoản của bạn đang được tạo",
    en: "Your profile is being created"
  },
  edit: { vi: "Chỉnh sửa", en: "Edit" },
  fullname: { vi: "Họ tên", en: "Fullname" },
  avatar: { vi: "Avatar", en: "Avatar" },
  back: { vi: "Quay lại", en: "Back" },
  loading: { vi: "Đang lưu", en: "Saving " },
  saved: { vi: "Đã lưu", en: "Saved" }
};

export default {
  name: "Profile",
  data: () => ({
    defaultAvatar: process.env.VUE_APP_DEFAULT_AVATAR,
    editMode: false,
    prestine: true
  }),
  computed: {
    ...mapGetters({
      language: "layout/language",
      authUser: "auth/authUser",
      profile: "auth/profile",
      loading: "auth/loading",
      error: "auth/error"
    }),
    $t() {
      return this.$translate(languagesMap, this.language.value);
    },
    fullname: {
      get: function() {
        return this.profile.fullname;
      },
      set: debounce(function(value) {
        if (value !== this.profile.fullname) {
          this.prestine = false;
          this.$store.dispatch("auth/updateProfile", { fullname: value });
        }
      }, 500)
    },
    avatar: {
      get: function() {
        return this.profile.avatar;
      },
      set: debounce(function(value) {
        if (value !== this.profile.avatar) {
          this.prestine = false;
          this.$store.dispatch("auth/updateProfile", { avatar: value });
        }
      }, 500)
    }
  },
  methods: {
    back() {
      this.editMode = false;
      this.prestine = true;
    }
  }
};
</script>
