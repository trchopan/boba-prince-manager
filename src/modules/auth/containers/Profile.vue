<template>
  <v-layout column>
    <my-header>{{ $t.greeting }}</my-header>
    <v-layout row wrap align-center justify-center>
      <v-flex xs12 sm6 md4>
        <transition name="fade" mode="out-in">
          <v-card v-if="!editMode"
            key="profile-display"
            class="elevation-1">
            <v-list>
              <v-list-tile avatar>
                <v-list-tile-avatar>
                  <img :src="profile.avatar" alt="big avatar">
                </v-list-tile-avatar>
                <v-list-tile-content>
                  <v-list-tile-title>{{ profile.fullname }}</v-list-tile-title>
                  <v-list-tile-sub-title>{{ profile.position | titleCase }}</v-list-tile-sub-title>
                </v-list-tile-content>
                <v-list-tile-action>
                  <v-btn
                    outline
                    @click="editMode = true">
                    {{ $t.edit }}
                  </v-btn>
                </v-list-tile-action>
              </v-list-tile>
            </v-list>
          </v-card>
          <v-card v-if="editMode"
            key="profile-edit"
            class="elevation-6 pt-3 px-3 text-xs-center">
            <v-avatar size="80">
              <img :src="profile.avatar" alt="big avatar">
            </v-avatar>
              <v-text-field
                id="fullname"
                prepend-icon="person"
                :label="$t.fullname"
                type="text"
                v-model="fullname"
                required />
              <v-text-field
                id="avatar"
                prepend-icon="lock"
                :label="$t.avatar"
                type="text"
                v-model="avatar"
                required />
            <v-card-actions>
              <v-spacer />
              <template v-if="!prestine">
                <v-icon class="mr-1" color="success">
                  {{ loading ? 'hourglass_empty' : 'check_circle_outline' }}
                </v-icon>
                <span class="mr-2" >{{ loading ? $t.loading : $t.done }}</span>
              </template>
              <v-btn
                color="primary"
                type="button"
                :disabled="loading"
                @click="back()">
                {{ $t.back }}
              </v-btn>
            </v-card-actions>
          </v-card>
        </transition>
      </v-flex>
    </v-layout>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import { debounce } from "@/helpers";

const languagesMap = {
  greeting: { vi: "Xin chào, bạn phẻ hok?", en: "Hi, how are you today?" },
  edit: { vi: "Chỉnh sửa", en: "Edit" },
  fullname: { vi: "Họ tên", en: "Fullname" },
  avatar: { vi: "Avatar", en: "Avatar" },
  back: { vi: "Quay lại", en: "Back" },
  loading: { vi: "Đang lưu", en: "Saving " },
  done: { vi: "Hoàn tất", en: "Done" }
};

export default {
  name: "Profile",
  data: () => ({
    editMode: false,
    prestine: true
  }),
  computed: {
    ...mapGetters({
      language: "layout/language",
      profile: "auth/profile",
      loading: "auth/loading",
      error: "auth/error"
    }),
    $t: this.$helpers.translate(languagesMap, this.language.code),
    fullname: {
      get: function() {
        return this.profile.fullname;
      },
      set: debounce(function(value) {
        if (value !== this.profile.fullname) {
          this.prestine = false;
          this.$store.dispatch("auth/updateProfile", { fullname: value });
        }
      }, 500)
    },
    avatar: {
      get: function() {
        return this.profile.avatar;
      },
      set: debounce(function(value) {
        if (value !== this.profile.avatar) {
          this.prestine = false;
          this.$store.dispatch("auth/updateProfile", { avatar: value });
        }
      }, 500)
    }
  },
  methods: {
    back() {
      this.editMode = false;
      this.prestine = true;
    }
  }
};
</script>
