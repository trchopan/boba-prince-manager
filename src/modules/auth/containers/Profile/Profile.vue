<template>
  <v-layout column>
    <my-header>{{ $t.greeting }}</my-header>
    <v-layout
      row
      align-center
      justify-center
    >
      <v-flex
        v-if="!profile"
        xs12
        sm8
        md6
      >
        <p class="text-xs-center">{{ $t.loadingProfile }}</p>
      </v-flex>
      <v-flex
        v-else
        xs12
        sm8
        md5
      >
        <transition
          name="fade"
          mode="out-in"
        >
          <v-layout
            v-if="!editMode"
            key="profile-display"
            column
            justify-center
          >
            <v-card class="elevation-1">
              <v-list>
                <v-list-tile avatar>
                  <v-list-tile-avatar>
                    <img
                      :src="profile.avatar"
                      alt="big avatar"
                    >
                  </v-list-tile-avatar>
                  <v-list-tile-content>
                    <v-list-tile-title>{{ profile.fullname }}</v-list-tile-title>
                    <v-list-tile-sub-title>{{ authUser.email }}</v-list-tile-sub-title>
                  </v-list-tile-content>
                  <v-list-tile-action>
                    <v-btn
                      outline
                      @click="editMode = true"
                    >{{ $t.edit }}</v-btn>
                  </v-list-tile-action>
                </v-list-tile>
              </v-list>
            </v-card>
            <v-card class="mt-3">
              <v-list two-line>
                <v-list-tile>
                  <v-list-tile-content>hi</v-list-tile-content>
                </v-list-tile>
              </v-list>
            </v-card>
          </v-layout>
          <ProfileEdit
            v-if="editMode"
            key="profile-edit"
            @back="back()"
          />
        </transition>
      </v-flex>
    </v-layout>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import ProfileEdit from "./ProfileEdit.vue";

const LANGUAGES_MAP = {
  greeting: { vi: "Xin chào, bạn phẻ hok?", en: "Hi, how are you today?" },
  loadingProfile: {
    vi: "Đang tải thông tin tài khoản",
    en: "Loading profile information"
  },
  edit: { vi: "Chỉnh sửa", en: "Edit" },
  permissions: { vi: "Quyền hạn", en: "Permissions" }
};

export default {
  name: "Profile",
  components: {
    ProfileEdit
  },
  data: () => ({
    defaultAvatar: process.env.VUE_APP_DEFAULT_AVATAR,
    editMode: false,
    prestine: true
  }),
  computed: {
    ...mapGetters({
      language: "layout/language",
      authUser: "auth/authUser",
      profile: "auth/profile",
      loading: "auth/loading"
    }),
    $t() {
      return this.$translate(LANGUAGES_MAP, this.language.value);
    }
  },
  methods: {
    back() {
      this.editMode = false;
      this.prestine = true;
    }
  }
};
</script>
