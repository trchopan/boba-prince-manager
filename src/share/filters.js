import Vue from "vue";

export const filters = {
  titleCase: str => {
    if (typeof str !== "string") {
      return "";
    }
    return str
      .toLowerCase()
      .split(" ")
      .map(function(word) {
        return word.replace(word[0], word[0].toUpperCase());
      })
      .join(" ");
  },
  format: (str, inputs) => {
    if (typeof str !== "string" || inputs.length === undefined) {
      return "";
    }
    let result = str;
    inputs.forEach(
      (input, index) => (result = result.replace(`{${index}}`, input))
    );
    return result;
  }
};

Object.keys(filters).forEach(key => Vue.filter(key, filters[key]));
