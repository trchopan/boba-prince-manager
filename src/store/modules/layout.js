import { logger } from "@/helpers";

const log = logger("[layout]");

export const themes = {
  light: "Light",
  dark: "Dark"
};

const state = {
  theme: themes.light,
  drawerOpen: false
};

const getters = {
  theme: state => state.theme,
  darkTheme: state => (state.theme === themes.dark ? true : false)
};

const actions = () => {
  const initLocalStorage = ({ commit }) => {
    if (window.localStorage !== undefined) {
      const theme = localStorage.theme;
      if (theme) {
        commit("themeChanged", theme);
      }
    } else {
      log("No Web Storage support");
    }
  };
  const changeTheme = ({ commit }, theme) => commit("themeChanged", theme);
  const toggleDrawer = ({ commit }) => commit("drawerToggled");
  return Object.freeze({ initLocalStorage, changeTheme, toggleDrawer });
};

const mutations = {
  themeChanged(state, theme) {
    state.theme = theme;
    localStorage.setItem("theme", theme);
    log("Theme changed", theme);
  },
  drawerToggled(state) {
    state.drawerOpen = !state.drawerOpen;
    log("Drawer toggled", state.drawerOpen);
  }
};

export default {
  namespaced: true,
  state,
  getters,
  actions: actions(),
  mutations
};
