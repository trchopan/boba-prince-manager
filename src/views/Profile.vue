<template>
  <v-layout column>
    <my-header>{{ $t.profile.greeting }}</my-header>
    <v-layout row wrap align-center justify-center>
      <v-flex xs12 sm6 md4>
        <v-list v-if="profile && !editMode" class="elevation-1">
          <v-list-tile avatar>
            <v-list-tile-avatar>
              <img :src="profile.avatar" alt="big avatar">
            </v-list-tile-avatar>
            <v-list-tile-content>
              <v-list-tile-title>{{ profile.fullname }}</v-list-tile-title>
              <v-list-tile-sub-title>{{ profile.position | titleCase }}</v-list-tile-sub-title>
            </v-list-tile-content>
            <v-list-tile-action>
              <v-btn
                outline
                @click="editMode = true">
                {{ $t.profile.edit }}
              </v-btn>
            </v-list-tile-action>
          </v-list-tile>
        </v-list>
        <v-card class="elevation-6 pt-3 px-3 text-xs-center" v-if="profile && editMode">
          <v-avatar size="80">
            <img :src="profile.avatar" alt="big avatar">
          </v-avatar>
          <v-text-field
            id="fullname"
            prepend-icon="person"
            name="fullname"
            :label="$t.profile.fullname"
            type="text"
            :error-messages="error ? error.message : ''"
            v-model="fullname" />
          <v-text-field
            id="avatar"
            prepend-icon="lock"
            name="avatar"
            :label="$t.profile.avatar"
            type="text"
            v-model="avatar" />
          <v-card-actions>
            <v-spacer />
            <v-icon class="mr-3" color="success">
              {{ loading ? 'hourglass_empty' : 'check_circle_outline' }}
            </v-icon>
            <v-btn
              color="primary"
              type="button"
              :disabled="loading"
              @click="editMode = false">
              {{ $t.profile.back }}
            </v-btn>
          </v-card-actions>
        </v-card>
      </v-flex>
    </v-layout>
  </v-layout>
</template>

<script>
import { mapGetters } from "vuex";
import { debounce } from "@/helpers";

export default {
  name: "Profile",
  data: () => ({
    editMode: false
  }),
  computed: {
    ...mapGetters({
      $t: "layout/$t",
      profile: "auth/profile",
      loading: "auth/loading",
      error: "auth/error"
    }),
    fullname: {
      get: function() {
        return this.profile.fullname;
      },
      set: debounce(function(value) {
        if (value !== this.profile.fullname) {
          this.$store.dispatch("auth/updateProfile", { fullname: value });
        }
      }, 500)
    },
    avatar: {
      get: function() {
        return this.profile.avatar;
      },
      set: debounce(function(value) {
        if (value !== this.profile.avatar) {
          this.$store.dispatch("auth/updateProfile", { avatar: value });
        }
      }, 500)
    }
  }
};
</script>
