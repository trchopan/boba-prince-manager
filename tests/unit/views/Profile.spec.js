import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Profile from "@/views/Profile.vue";
import { mockCustomElements } from "@tests/unit/__mocks__/custom-elements.js";
import { docData } from "@tests/unit/__mocks__/firebase-results.js";
import { languages } from "@/store/modules/layout.js";
import { filters } from "@/filters.js";

const localVue = createLocalVue();
localVue.filter("titleCase", filters.titleCase);
localVue.use(Vuex);

describe("views/Profile.vue", () => {
  let store;
  let getters;
  let actions;
  let wrapper;

  beforeEach(() => {
    getters = {
      "layout/language": () => languages.en,
      "auth/profile": () => docData,
      "auth/error": () => null,
      "auth/loading": () => false
    };
    actions = {
      "auth/updateProfile": jest.fn()
    };
    store = new Vuex.Store({
      getters,
      actions
    });
    wrapper = shallowMount(Profile, {
      stubs: mockCustomElements,
      store,
      localVue
    });
  });
  it("has all the tools", () => {
    expect(wrapper.vm.editMode).toBeDefined();
  });
  it("changes user profile when input change with debouncing", done => {
    const otherValue = "Other Value";
    wrapper.vm.fullname = "afiauhnilg";
    wrapper.vm.avatar = "wertniwertugn";
    wrapper.vm.fullname = otherValue;
    wrapper.vm.avatar = otherValue;

    setTimeout(() => {
      expect(actions["auth/updateProfile"].mock.calls[0][1]).toEqual({
        fullname: otherValue
      });
      expect(actions["auth/updateProfile"].mock.calls[1][1]).toEqual({
        avatar: otherValue
      });
      done();
    }, 1000);
  });
  it("matchs snapshot", () => {
    expect(wrapper).toMatchSnapshot();
    wrapper.vm.editMode = true;
    expect(wrapper).toMatchSnapshot();
  });
});
