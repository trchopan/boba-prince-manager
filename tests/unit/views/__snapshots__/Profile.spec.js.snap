// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/Profile.vue matchs snapshot 1`] = `
<v-layout-stub column="">
  <my-header-stub>Hi there, how are you today?</my-header-stub>
  <v-layout-stub row="" wrap="" align-center="" justify-center="">
    <v-flex-stub xs12="" sm6="" md4="">
      <v-list-stub class="elevation-1">
        <v-list-tile-stub avatar="">
          <v-list-tile-avatar-stub><img src="avatar" alt="big avatar"></v-list-tile-avatar-stub>
          <v-list-tile-content-stub>
            <v-list-tile-title-stub>Tester</v-list-tile-title-stub>
            <v-list-tile-sub-title-stub>Manager</v-list-tile-sub-title-stub>
          </v-list-tile-content-stub>
          <v-list-tile-action-stub>
            <v-btn-stub outline="">
              Edit
            </v-btn-stub>
          </v-list-tile-action-stub>
        </v-list-tile-stub>
      </v-list-stub>
      <!---->
    </v-flex-stub>
  </v-layout-stub>
</v-layout-stub>
`;

exports[`views/Profile.vue matchs snapshot 2`] = `
<v-layout-stub column="">
  <my-header-stub>Hi there, how are you today?</my-header-stub>
  <v-layout-stub row="" wrap="" align-center="" justify-center="">
    <v-flex-stub xs12="" sm6="" md4="">
      <!---->
      <v-card-stub class="elevation-6 pt-3 px-3 text-xs-center">
        <v-avatar-stub size="80"><img src="avatar" alt="big avatar"></v-avatar-stub>
        <v-text-field-stub id="fullname" prepend-icon="person" name="fullname" label="Fullname" type="text" error-messages=""></v-text-field-stub>
        <v-text-field-stub id="avatar" prepend-icon="lock" name="avatar" label="Avatar" type="text"></v-text-field-stub>
        <v-card-actions-stub>
          <v-spacer-stub></v-spacer-stub>
          <v-icon-stub color="success" class="mr-3">
            check_circle_outline
          </v-icon-stub>
          <v-btn-stub color="primary" type="button">
            Back
          </v-btn-stub>
        </v-card-actions-stub>
      </v-card-stub>
    </v-flex-stub>
  </v-layout-stub>
</v-layout-stub>
`;
